### DISCUSSION TITLE
`keepalive` with body being async generator

### DISCUSSION THREAD

**Discussion by brettwillis** [nodejs/undici#2372] [Q&A]
https://github.com/nodejs/undici/blob/8050ec0224a51d44f776364820e6a16112fb4781/lib/fetch/body.js#L189-L193

It seems that if `RequestInit.body` is an async generator, then `RequestInit.keepalive` is not allowed and we get an error. Why is this? It makes it hard to use `keepalive` without knowing ahead of time what type of body will be given.

Also `keepalive` is not documented here at all... what does it do in `undici`'s implementation of fetch? I'm guessing it has nothing to do with keeping the TCP connection/socket alive between requests?

**Comment by KhafraDev** [comment]
It's part of the fetch spec, I'm sure someone can give you an answer if you asked on their [GitHub](https://github.com/whatwg/fetch) or their [chat](https://whatwg.org/chat). Note that the spec doesn't typically allow *any* async iterable, so this is part of the spec dealing with how ReadableStreams are handled.

**Comment by brettwillis** [comment]
Ok I think I can answer my own question: `Request.keepalive` does nothing in the context of `undici`:

https://github.com/nodejs/undici/blob/a3a437eeebeb8fcca9014107b9ee748402ba2b1e/lib/fetch/index.js#L1330-L1334

That is *nothing except possibly throw more errors*! I guess this error related to async generators is to remain semantically true to the spec that the request could not be sent if the JavaScript context that the generator function runs inside was no longer around? However I don't see how that differs from any other type of body, seeing as it is possible for `ReadableStream` to wrap a generator function anyways, and this is all irrelevant if the Node.js process has exited...

From MDN, [`keepalive`](https://developer.mozilla.org/en-US/docs/Web/API/fetch#keepalive):

> The keepalive option can be used to allow the request to outlive the page. Fetch with the keepalive flag is a replacement for the `Navigator.sendBeacon()` API.

Correct me if I'm wrong, but an API whose only purpose is to throw an error in an edge-case of body type seems like a bit of a trap. I'd prefer to see it ignored completely rather than throw an error in this case don't you think?

Furthermore it would be beneficial to have some documenation in the type definitions stating that `keepalive` is a no-op.
